File: tests/streaming/perf_bench/perf_bench.ts
// perf_bench.ts — RN/Web consumer for the GAP 09 #8 perf bench.
//
// v3.2: in-process decode latency (mirrors perf_bench.kt / perf_bench.dart).
//
// Why not `now() - metrics.created_at_ns`: the C++ producer stamps
// `created_at_ns` from `std::chrono::steady_clock`, which is monotonic
// but has a process-local, platform-defined epoch. Node's
// `process.hrtime.bigint()` is also monotonic with its own origin
// (process start). Subtracting the two yields garbage deltas (observed:
// ~158 s offset on macOS), which the `>0` filter still accepts but
// blows past the 1ms p50 budget. Decode latency in-process is what the
// README spec actually describes — "per-event work: proto decode +
// delta computation only" — so we bracket `process.hrtime.bigint()`
// directly around the `decode` callback. Same clock, same process,
// measures exactly the cost the p50 budget is meant to bound.
//
// Single source for both React Native (Jest) and Web (Vitest) SDKs.
//
// Input format (matches perf_producer.cpp):
//   uint32_t magic = 0x42504152 ('RAPB')
//   uint32_t count
//   count × { uint32_t len; uint8_t[len] proto_bytes }

import * as fs from 'fs';

const DEFAULT_INPUT_PATH = '/tmp/perf_input.bin';
const MAGIC = 0x42504152;

/**
 * Decode one VoiceEvent frame and return the producer-side
 * `metrics.created_at_ns` if present, else null. Used only to
 * count "non-empty" frames; the perf delta is the bracketed
 * decode time, not a cross-clock subtraction.
 */
export type VoiceEventDecoder = (frame: Uint8Array) => bigint | null;

/**
 * Run the perf bench consumer against /tmp/perf_input.bin. Each
 * delta is the in-process decode time of one VoiceEvent (bracketed
 * with `process.hrtime.bigint()`), not a cross-process clock diff.
 *
 * @returns { count, nonEmpty } — `nonEmpty` counts frames whose
 *           MetricsEvent arm carries `created_at_ns > 0`.
 */
export async function runPerfBench(
  outputPath: string,
  decode: VoiceEventDecoder,
  inputPath: string = DEFAULT_INPUT_PATH,
): Promise<{ count: number; nonEmpty: number }> {
  const buf = fs.readFileSync(inputPath);
  const view = new DataView(buf.buffer, buf.byteOffset, buf.byteLength);

  if (buf.length < 8) {
    throw new Error('perf_bench input too short (<8 bytes)');
  }
  const magic = view.getUint32(0, true);
  if (magic !== MAGIC) {
    throw new Error(
      `perf_bench bad magic: 0x${magic.toString(16)} (expected 0x${MAGIC.toString(16)})`,
    );
  }
  const count = view.getUint32(4, true);
  const deltas = new BigInt64Array(count);

  let cursor = 8;
  let nonEmpty = 0;

  for (let i = 0; i < count; i++) {
    if (cursor + 4 > buf.length) break;
    const len = view.getUint32(cursor, true);
    cursor += 4;
    if (cursor + len > buf.length) break;

    const frame = buf.subarray(cursor, cursor + len);
    cursor += len;

    // Bracket decode in-process — same monotonic clock, same process.
    const startNs = process.hrtime.bigint();
    const producerNs = decode(frame);
    const endNs = process.hrtime.bigint();
    const decodeNs = endNs - startNs;

    if (decodeNs > 0n) {
      deltas[i] = decodeNs;
      if (producerNs !== null && producerNs > 0n) nonEmpty++;
    } else {
      deltas[i] = 0n;
    }
  }

  const lines = Array.from(deltas).map((d) => d.toString()).join('\n') + '\n';
  fs.writeFileSync(outputPath, lines);

  // eslint-disable-next-line no-console
  console.log(
    `perf_bench.ts: wrote ${count} deltas (${nonEmpty} non-empty) to ${outputPath}`,
  );
  return { count, nonEmpty };
}

File: tests/streaming/perf_bench/perf_bench.kt
/**
 * perf_bench.kt — Kotlin consumer for the GAP 09 #8 perf bench.
 *
 * v3.2: Reads /tmp/perf_input.bin (produced by
 * tests/streaming/perf_bench/perf_producer.cpp), decodes each VoiceEvent
 * via Wire (the Kotlin proto lib used by the SDK), and measures the
 * per-event decode latency with `System.nanoTime()` brackets around
 * the decode call. Writes per-event delta_ns to /tmp/perf_bench.kt.log.
 *
 * Why not `now() - metrics.created_at_ns` like the C++/Swift consumers:
 * the producer stamps `created_at_ns` using C++ `std::chrono::steady_clock`,
 * which is monotonic but has a process-local, platform-defined epoch.
 * The JVM's `System.nanoTime()` is also monotonic but has its own
 * origin; on macOS it's backed by `mach_absolute_time`, on Linux by
 * `CLOCK_MONOTONIC` — neither guaranteed to share an epoch with the
 * C++ producer running in a separate process. Subtracting them yields
 * garbage deltas (observed: all-negative ≈ -96s offset), which the
 * `p50 > 0` filter discards, producing a spurious "no non-zero deltas"
 * failure. Measuring decode latency in-process is what the README spec
 * actually describes — "per-event work: proto decode + delta
 * computation only" — so we bracket `System.nanoTime()` directly
 * around the decode call. Monotonic, same-clock, measures exactly
 * the cost the p50 budget is meant to bound.
 *
 * Runner integration: sdk/runanywhere-kotlin/src/jvmTest/kotlin/
 * PerfBenchTest.kt (JUnit wrapper — asserts p50 < 1ms).
 *
 * Binary format (matches perf_producer.cpp):
 *   uint32_t magic = 0x42504152 ('RAPB')
 *   uint32_t count
 *   count × { uint32_t len; uint8_t[len] proto_bytes }
 */

package tests.streaming.perf_bench

import ai.runanywhere.proto.v1.VoiceEvent
import java.io.File
import java.nio.ByteBuffer
import java.nio.ByteOrder

object PerfBench {
    const val DEFAULT_INPUT_PATH = "/tmp/perf_input.bin"
    const val DEFAULT_OUTPUT_PATH = "/tmp/perf_bench.kt.log"
    const val MAGIC = 0x42504152u  // 'RAPB'

    data class Result(
        val count: Int,
        val nonEmpty: Int,
        val deltas: List<Long>,
    )

    /**
     * Run the perf bench consumer. Returns per-event latency deltas in ns.
     */
    fun run(
        inputPath: String = DEFAULT_INPUT_PATH,
        outputPath: String = DEFAULT_OUTPUT_PATH,
    ): Result {
        val bytes = File(inputPath).readBytes()
        require(bytes.size >= 8) { "perf_bench input too short (<8 bytes): ${bytes.size}" }

        val header = ByteBuffer.wrap(bytes, 0, 8).order(ByteOrder.LITTLE_ENDIAN)
        val readMagic = header.int.toUInt()
        require(readMagic == MAGIC) {
            "perf_bench bad magic: 0x${readMagic.toString(16)} (expected 0x${MAGIC.toString(16)})"
        }
        val count = header.int

        val deltas = ArrayList<Long>(count)
        var nonEmpty = 0
        var cursor = 8

        for (i in 0 until count) {
            if (cursor + 4 > bytes.size) break
            val lenBuf = ByteBuffer.wrap(bytes, cursor, 4).order(ByteOrder.LITTLE_ENDIAN)
            val len = lenBuf.int
            cursor += 4
            if (cursor + len > bytes.size) break

            val frame = bytes.copyOfRange(cursor, cursor + len)
            cursor += len

            try {
                val startNs = System.nanoTime()
                val event = VoiceEvent.ADAPTER.decode(frame)
                val endNs = System.nanoTime()
                val decodeNs = endNs - startNs

                if (decodeNs > 0L) {
                    deltas.add(decodeNs)
                    if ((event.metrics?.created_at_ns ?: 0L) > 0L) {
                        nonEmpty++
                    }
                } else {
                    deltas.add(0L)
                }
            } catch (e: Exception) {
                deltas.add(0L)
            }
        }

        writeDeltas(deltas, outputPath)
        println("perf_bench.kt: wrote ${deltas.size} deltas ($nonEmpty non-empty) to $outputPath")
        return Result(count = deltas.size, nonEmpty = nonEmpty, deltas = deltas)
    }

    /** Compute p50 over non-zero deltas. Returns null if no non-zero values. */
    fun p50(deltas: List<Long>): Long? {
        val nonZero = deltas.filter { it > 0 }.sorted()
        if (nonZero.isEmpty()) return null
        return nonZero[nonZero.size / 2]
    }

    private fun writeDeltas(deltas: List<Long>, path: String) {
        File(path).writeText(deltas.joinToString("\n") + "\n")
    }
}
